### PR TITLE
 [wip] FWaaS v2.0 Rule List

### PR DESCRIPTION
**Pull request by Elethiomel**
For #514  

* https://github.com/openstack/neutron-fwaas/blob/27906d0acff349c9e5ac7955d203fb2233c295db/neutron_fwaas/db/firewall/v2/firewall_db_v2.py#L68-L82 
* https://github.com/openstack/neutron-lib/blob/56033ba643812a30577f6ab17648806c2ee494ba/neutron_lib/api/definitions/firewall_v2.py#L47-L114

### CODE REVIEW

**Comment by Elethiomel** [comment]
recheck

**Comment by Elethiomel** [comment]
recheck

**Comment by Elethiomel** [comment]
@jtopjian Openlab checks failing. Doesn't seem to be related to my code though :)

**Comment by Elethiomel** [comment]
@jtopjian There we go. Recheck successful :)

File: openstack/networking/v2/extensions/fwaas_v2/rules/requests.go [hunk 1]
```diff
@@ -39,6 +40,74 @@ const (
...
+	Name                 string   `q:"name"`
+	Description          string   `q:"description"`
+	Protocol             Protocol `q:"protocol"`
+	Action               Action   `q:"action"`
+	IPVersion            int      `q:"ip_version"`
+	SourceIPAddress      string   `q:"source_ip_address"`
+	DestinationIPAddress string   `q:"destination_ip_address"`
+	SourcePort           string   `q:"source_port"`
+	DestinationPort      string   `q:"destination_port"`
+	Enabled              bool     `q:"enabled"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/fwaas_v2/rules/requests.go]: By default, Gophercloud does not set zero-values in the query strings. Using `bool`, we'll only ever be able to query `?enabled=true`. We should change this to `*bool` so we can do `?enabled=false`.

File: openstack/networking/v2/extensions/fwaas_v2/rules/requests.go [hunk 2]
```diff
@@ -39,6 +40,74 @@ const (
...
+	ID                   string   `q:"id"`
+	Shared               bool     `q:"shared"`
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/fwaas_v2/rules/requests.go]: Same here

File: openstack/networking/v2/extensions/fwaas_v2/rules/requests.go [hunk 3]
```diff
@@ -90,6 +159,47 @@ func Get(c *gophercloud.ServiceClient, id string) (r GetResult) {
 	return
 }
 
+// UpdateOptsBuilder is the interface options structs have to satisfy in order
```
Comment (jtopjian) [inline openstack/networking/v2/extensions/fwaas_v2/rules/requests.go]: I think this PR might need rebased off of Gophercloud's master branch. Update has already been merged: https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/extensions/fwaas_v2/rules/requests.go#L93

Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/rules/requests.go]: Whoops :D Good catch, thanks!

File: openstack/networking/v2/extensions/fwaas_v2/rules/requests.go (hunk 1 above)
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/rules/requests.go]: Should be resolved in commit 4e8d6bc7085d63aea8479da88fba68c6a2a985a6

File: openstack/networking/v2/extensions/fwaas_v2/rules/requests.go (hunk 2 above)
Comment (Elethiomel) [reply openstack/networking/v2/extensions/fwaas_v2/rules/requests.go]: Should be resolved in commit 4e8d6bc7085d63aea8479da88fba68c6a2a985a6

**Comment by Elethiomel** [comment]
@jtopjian Look OK to you?

**Comment by jtopjian** [comment]
@Elethiomel I apologize for the delay. I will review this as soon as I get a chance.

**Comment by jtopjian** [comment]
@Elethiomel It looks like you've got a good understanding of how to implement this, including understanding how it relates to the Python-based service code.

Please feel free to combine CRUD into a single PR for the rest of the resources.

**Pull request merged**